x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportDeleteTrainedModelAction.java

```diff
@@ -0,0 +1,134 @@
+
+/**
+ * The action is a master node action to ensure it reads an up-to-date cluster
+ * state in order to determine whether there is a persistent task for the analytics
+ * to delete.
+ */
+public class TransportDeleteTrainedModelAction
+    extends TransportMasterNodeAction<DeleteTrainedModelAction.Request, AcknowledgedResponse> {
+
+    private static final Logger LOGGER = LogManager.getLogger(TransportDeleteTrainedModelAction.class);
+
+    private final TrainedModelProvider trainedModelProvider;
+    private final InferenceAuditor auditor;
+    private final IngestService ingestService;
+
+    @Inject
+    public TransportDeleteTrainedModelAction(TransportService transportService, ClusterService clusterService,
+                                             ThreadPool threadPool, ActionFilters actionFilters,
+                                             IndexNameExpressionResolver indexNameExpressionResolver,
+                                             TrainedModelProvider configProvider, InferenceAuditor auditor,
+                                             IngestService ingestService) {
+        super(DeleteTrainedModelAction.NAME, transportService, clusterService, threadPool, actionFilters,
+            DeleteTrainedModelAction.Request::new, indexNameExpressionResolver);
+        this.trainedModelProvider = configProvider;
+        this.ingestService = ingestService;
+        this.auditor = Objects.requireNonNull(auditor);
+    }
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.SAME;
+    }
+
+    @Override
+    protected AcknowledgedResponse read(StreamInput in) throws IOException {
+        return new AcknowledgedResponse(in);
+    }
+
+    @Override
+    protected void masterOperation(Task task,
+                                   DeleteTrainedModelAction.Request request,
+                                   ClusterState state,
+                                   ActionListener<AcknowledgedResponse> listener) {
+        String id = request.getId();
+        IngestMetadata currentIngestMetadata = state.metaData().custom(IngestMetadata.TYPE);
+        Set<String> referencedModels = getReferencedModelKeys(currentIngestMetadata);
+
+        if (referencedModels.contains(id)) {
+            listener.onFailure(new ElasticsearchStatusException("Cannot delete mode [{}] as it is still referenced by ingest processors",
+                RestStatus.CONFLICT,
+                id));
+            return;
+        }
+
+        trainedModelProvider.deleteTrainedModel(request.getId(), ActionListener.wrap(
+            r -> {
+                auditor.info(request.getId(), "trained model deleted");
+                listener.onResponse(new AcknowledgedResponse(true));
+            },
+            listener::onFailure
+        ));
+    }
+
+    private Set<String> getReferencedModelKeys(IngestMetadata ingestMetadata) {
+        Set<String> allReferencedModelKeys = new HashSet<>();
+        if (ingestMetadata != null) {
```

Comment:
This is a matter of taste, but I would add this in the beginning of this method:
```
if (ingestMetadata == null) {
  return Collections.emptySet();
}
```

This way it is clear what happens on `null` case and the rest of the method focuses on `non-null` case.